### DISCUSSION TITLE
Integrating Stripe with Nhost

### DISCUSSION THREAD

**Discussion by nunopato** [nhost/nhost#772] [Ideas]
Hey, 

we are looking into how we could better integrate Stripe with Nhost and would love to get your input to understand how are you all using both together. 

## Features

What features are you mostly using and would like to see available somehow? some examples:

- list/create customers
- list/create customer subscriptions
- list/create invoices
- list/create products
- etc

## APIs and Data

We're thinking about reading and writing data from / to stripe differently and will start working on reading.

### Queries

Through remote schemas, we are able to expose stripe data through your graphql API:

```gql
query getUsers {
  users {  
    name
    email
    invoices { # from stripe
      amount
      date
    }
  }
}
```
As you can see, it is very easy to read data from stripe as part of other queries.

### Mutations

Mutations are better done server-side using nhost functions and here you have to use stripe's SDK. Let's imagine the scenario where an event trigger is executed when a new user is inserted and its corresponding stripe customer is created:

```ts
import Stripe from 'stripe';

const stripe = new Stripe(process.env.STRIPE_SK, {
  apiVersion: '2020-08-27',
});

const handler = async (req: Request, res: Response) => {
  const { body }: { body: HasuraBody } = req;
  
  const { id: stripeCustomerId } = await stripe.customers.create({
    name: body.event.data.new.name,
    email: body.event.data.new.email,
  });
  
  // do something with it 
  
  return res.send('OK');
}
``` 

We will work on code examples and documentation for mutations. Unfortunately, they are a bit trickier to expose through the API and that is why we are thinking only of working on querying now.

### Data

What Stripe data are you storing in your database and why? 

We see 2 options here:

1. We store `stripe_customer_id` and `stripe_subscription_id` and all data is fetched on the fly. No prior syncing between your database and your stripe account (this is how we do it at nhost). 
2. We create a `stripe` schema on your database with tables like `invoices`, `customers`, etc and a `/webhooks` endpoint that would populate those when an event is sent from Stripe (e.g., when a customer gets created). Do you see any use-cases here?

The `/webhooks` endpoint could be valuable for other workflows as well. Let's imagine a credit card that gets broken (for some irrelevant reason), if a webhook is defined for when a payment fails, we could have some logic taking care of that.

Your feedback is much appreciated, so please let us know how you are using stripe and help us shape its integration.

**Comment by TaylorFacen** [comment]
Having a pre-built remote schema would be great. For my app, I currently use Hasura actions to query stripe data in GraphQL format along with nhost functions for tasks like creating a billing portal session. My current workflow is:

1. When a user signs up for my app, a database event function triggers (`on_insert_user`)
2. In this function I check to see if the user already has a stripe customer id (maybe they're signing up again). If not, I create a stripe customer for them
3. I then grab their last subscription if they have one or create a new subscription with a 14 day trial. 
4. I store their stripe customer Id in the metadata column of their user record

Whenever someone logs into the app, I use their stripe customer id to fetch their latest subscription using a hasura action. 


Back when I was using Firebase, I *really* liked their [stripe integration](https://firebase.google.com/products/extensions/stripe-firestore-stripe-payments).

**Reply by nunopato** [reply]
So all you're storing is `stripe_customer_id`. Seems pretty straightforward :+1: thank you!

Anything, in particular, you liked about the Firebase integration?

**Comment by cbsmith402** [comment]
Another option for implementing #2 above is to use the `/events` endpoint to get everything that happened since the last sync, rather than using webhooks, like [tabledog](https://github.com/tabledog/tdog-cli) does. That way there's less of a chance of missing data somehow.
